File: Aliucord/src/main/java/com/aliucord/api/ButtonsAPI.java
package com.aliucord.api;

import android.view.View;

import com.aliucord.utils.ReflectUtils;
import com.discord.api.botuikit.*;
import com.discord.models.message.Message;
import com.discord.stores.StoreStream;

import java.util.*;

public class ButtonsAPI {
    private static final Map<Integer, View.OnClickListener> listeners = new HashMap<>();
    private static int lastListenerID = 0;

    /**
     * Adds ButtonComponent to Message object
     *
     * @param message         Message you want to add button.
     * @param buttonComponent ButtonComponent object, You can create it using ButtonsAPI.createButton().
     */
    public static void addButton(Message message, ButtonComponent buttonComponent) {
        var components = message.getComponents();
        if (components == null) {
            components = new ArrayList<>();
            try {
                ReflectUtils.setField(message, "components", components);
            } catch (NoSuchFieldException | IllegalAccessException e) {
                // DONT CARE;
            }
        }
        components.add(buttonComponent);
        StoreStream.access$getDispatcher$p(StoreStream.getPresences().getStream()).schedule(() -> {
            StoreStream.access$handleMessageUpdate(StoreStream.getPresences().getStream(), message.synthesizeApiMessage());
            return null;
        });
    }

    /**
     * Creates ButtonComponent object, To use this call ButtonsAPI.addButton()
     *
     * @param label           Label of the Button.
     * @param buttonStyle     Type of the Button.
     * @param onClickListener Listener that will get called when button is clicked
     * @return ButtonComponent object, if some error happens returns null
     */
    public static ButtonComponent createButton(String label, ButtonStyle buttonStyle, View.OnClickListener onClickListener) {
        try {
            var id = ButtonsAPI.generateListenerID(onClickListener);
            var component = ReflectUtils.allocateInstance(ButtonComponent.class);
            ReflectUtils.setField(component, "label", label);
            ReflectUtils.setField(component, "disabled", false);
            ReflectUtils.setField(component, "type", ComponentType.BUTTON);
            ReflectUtils.setField(component, "style", buttonStyle);
            ReflectUtils.setField(component, "url", "aliucord://" + id);
            return component;
        } catch (Exception e) {
            return null;
        }
    }

    private static synchronized int generateListenerID(View.OnClickListener listener) {
        listeners.put(lastListenerID, listener);
        return lastListenerID++;
    }

    /**
     * Calls listener with id
     *
     * @param id ID of the listener.
     * @param v  View that is clicked.
     */
    public static void callListenerWithID(int id, View v) {
        if (listeners.containsKey(id)) listeners.get(id).onClick(v);
    }
}

File: aliucord/src/main/java/com/aliucord/api/ButtonsAPI.kt
/*
 * Copyright (c) 2021 Juby210 & Vendicated
 * Licensed under the Open Software License version 3.0
 */

package com.aliucord.api

import android.content.Context
import androidx.fragment.app.FragmentActivity

import com.aliucord.coreplugins.ButtonsAPI
import com.aliucord.Logger
import com.aliucord.utils.ReflectUtils
import com.aliucord.utils.lazyField

import com.discord.api.botuikit.*
import com.discord.models.message.Message
import com.discord.stores.StoreStream

import java.util.*
import kotlin.collections.lastOrNull

/**
 * Adds methods for creating button components
 */
@Suppress("UNCHECKED_CAST")
object ButtonsAPI {

    /**
     * Stores data about a button
     */
    data class ButtonData(val label: String, val style: ButtonStyle, val onPress: (Message, FragmentActivity) -> Unit)

    private val componentsField by lazyField<ActionRowComponent>()
    private val msgComponentsField by lazyField<Message>("components")
    private val arTypeField by lazyField<ActionRowComponent>("type")

    private val labelField by lazyField<ButtonComponent>()
    private val styleField by lazyField<ButtonComponent>()
    private val disabledField by lazyField<ButtonComponent>()
    private val idField by lazyField<ButtonComponent>("customId")
    private val typeField by lazyField<ButtonComponent>()

    /**
     * Creates a button with the given data
     * @param button The data to create the button with
     */
    @JvmStatic
    fun Message.addButton(button: ButtonData) {
        this.addButton(button.label, button.style, button.onPress)
    }

    /**
     * Adds a button component to the message.
     * @param label   The label of the button.
     * @param style   The style of the button. {@link ButtonStyle}
     * @param onPress Callback for when the button is pressed, passing the message as an argument.
     */
    @JvmStatic
    fun Message.addButton(label: String, style: ButtonStyle, onPress: (Message, FragmentActivity) -> Unit) {
        val id = (-CommandsAPI.generateId()).toString()
        val components = this.components ?: ArrayList<Component>().also { components ->
            msgComponentsField[this] = components
        }

        try {
            val buttonComponent = ReflectUtils.allocateInstance(ButtonComponent::class.java) as ButtonComponent

            labelField[buttonComponent] = label
            styleField[buttonComponent] = style
            disabledField[buttonComponent] = false
            typeField[buttonComponent] = ComponentType.BUTTON
            idField[buttonComponent] = "${-CommandsAPI.ALIUCORD_APP_ID}-${id}"

            val row = components.lastOrNull() ?: ReflectUtils.allocateInstance<ActionRowComponent>(ActionRowComponent::class.java).also { row ->
                arTypeField[row] = ComponentType.ACTION_ROW
                components.add(row)
            }

            val rowItems = componentsField[row] as ArrayList<Component>? ?: ArrayList<Component>().also { rowItems ->
                componentsField[row] = rowItems
            }

            rowItems.add(buttonComponent)
            ButtonsAPI.actions[id] = onPress

            StoreStream.`access$getDispatcher$p`(StoreStream.getPresences().stream).schedule {
                StoreStream.`access$handleMessageUpdate`(StoreStream.getPresences().stream, synthesizeApiMessage())
            }
        } catch (t: Throwable) {
            Logger("ButtonsAPI").error("Failed to create button component", t)
        }
    }

}
